File: src/main/java/me/ksviety/teleporter/ShiftedVec3i.java
package me.ksviety.teleporter;

import net.minecraft.util.math.Vec3i;

public class ShiftedVec3i extends Vec3i {
    public static ShiftedVec3i shift(Vec3i original, int shift, ShiftAxis axis) {
        int x = original.getX();
        int y = original.getY();
        int z = original.getZ();

        if (axis == ShiftAxis.Forward)
            z += shift;
        else if (axis == ShiftAxis.Backward)
            z -= shift;
        else if (axis == ShiftAxis.Right)
            x += shift;
        else if (axis == ShiftAxis.Left)
            x -= shift;

        return new ShiftedVec3i(x, y, z);
    }

    private ShiftedVec3i(int x, int y, int z) {
        super(x, y, z);
    }
}

File: src/main/kotlin/me/ksviety/teleporter/ShiftedVec3i.kt
package me.ksviety.teleporter

import net.minecraft.util.math.Vec3i

class ShiftedVec3i private constructor(x: Int, y: Int, z: Int) : Vec3i(x, y, z) {
    companion object {

        fun shift(original: Vec3i, shift: Int, axis: ShiftAxis): ShiftedVec3i {
            fun Int.bidirectionalizedByAxis(positive: ShiftAxis, negative: ShiftAxis): Int = when (axis) {
                positive -> +this
                negative -> -this
                else -> 0
            }

            val x = original.x + shift.bidirectionalizedByAxis(ShiftAxis.Right, ShiftAxis.Left)
            val y = original.y + shift.bidirectionalizedByAxis(ShiftAxis.Up, ShiftAxis.Down)
            val z = original.z + shift.bidirectionalizedByAxis(ShiftAxis.Forward, ShiftAxis.Backward)

            return ShiftedVec3i(x, y, z)
        }
    }
}
